vaultclient/vaultclient.go

```
@@ -1,63 +1,175 @@
-	req := c.NewRequest(op, params, data)
+type APIError struct {
+	StatusCode int
+	Err        error
+}
+
+func (h *APIError) Error() string {
+	return h.Err.Error()
+}
+
+func (h *APIError) Unwrap() error {
+	return h.Err
+}
+
+func (h *APIError) HTTPStatusCode() int {
+	return h.StatusCode
+}
+
+func (vc *Vault) handleAWSResponse(resp *http.Response) (respBody []byte, err error) {
+	defer func() {
+		if closeErr := resp.Body.Close(); closeErr != nil && err == nil {
+			err = fmt.Errorf("failed to close response body: %w", closeErr)
+		}
+	}()
+
+	respBody, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response: %w", err)
+	}
+
+	if resp.StatusCode >= 400 {
+		var xmlError struct {
+			Error struct {
+				Code    string `xml:"Code"`
+				Message string `xml:"Message"`
+			} `xml:"Error"`
+			RequestID string `xml:"RequestId"`
+		}
+		if err := xml.Unmarshal(respBody, &xmlError); err == nil {
+			switch xmlError.Error.Code {
+			case "EntityAlreadyExists":
+				return nil, &APIError{
+					StatusCode: resp.StatusCode,
+					Err: &types.EntityAlreadyExistsException{
+						Message: &xmlError.Error.Message,
+					},
+				}
+			case "InvalidInput":
+				return nil, &APIError{
+					StatusCode: resp.StatusCode,
+					Err: &types.InvalidInputException{
+						Message: &xmlError.Error.Message,
+					},
+				}
+			case "NoSuchEntity":
+				return nil, &APIError{
+					StatusCode: resp.StatusCode,
+					Err: &types.NoSuchEntityException{
+						Message: &xmlError.Error.Message,
+					},
+				}
+			case "DeleteConflict":
+				return nil, &APIError{
+					StatusCode: resp.StatusCode,
+					Err: &types.DeleteConflictException{
+						Message: &xmlError.Error.Message,
+					},
+				}
+			default:
+				return nil, &APIError{
+					StatusCode: resp.StatusCode,
+					Err: &types.ServiceFailureException{
+						Message: &xmlError.Error.Message,
+					},
+				}
+			}
+		}
```

does this case happen often? is it really useful to make this case different from the next one, where we failed to parse xmml?

Either remove the useless `default`:
```suggestion
		}
```

or change the code to "fail fast" and avoid the large nested block:

```suggestion
if err := xml.Unmarshal(respBody, &xmlError); err != nil {
    message := string(respBody)
		return nil, &APIError{
			StatusCode: resp.StatusCode,
			Err: &types.ServiceFailureException{
				Message: &message,
			},
		}
}

switch xmlError {
    ...
default:
    ...
}
```

---

I suppose the default case can regularly happen as the switch is definitely not matching on all the different types of errors. I chanegd it to fail early